staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go

```diff
@@ -42,12 +42,13 @@ func SchemaDeclType(s *schema.Structural) *DeclType {
 		})
 	}
 	if s.XEmbeddedResource {
-		// TODO: load the schema for the type if it is in the system? Is it required to be?
 		return NewObjectType("embedded", map[string]*DeclField{
-			"kind": {Name: "kind", Type: StringType},
+			"kind":       {Name: "kind", Type: StringType},
 			"apiVersion": {Name: "apiVersion", Type: StringType},
-			"metadata": {Name: "metadata", Type: DynType, Required: false},
-			"spec": {Name: "spec", Type: DynType},
+			"metadata": {Name: "metadata", Type: NewObjectType("metadata", map[string]*DeclField{
+				"name": {Name: "name", Type: StringType},
+			})},
+			"spec":   {Name: "spec", Type: DynType},
 			"status": {Name: "status", Type: DynType, Required: false},
```

Comment:
hmm, these are strange. Spec and status are not something always existing. And they could even be of different type, e.g. `spec` of string type. Technically possible.

Reply:
DynType means the contents of the field can be anything and that all type checking is deferred to runtime.

That said, looks like there might be problems with what I've put here. Can there be other fields at the same level of spec? If so, my only option is to make the entire embedded object DynType.  I don't think it's possible to say it's _mostly_ dynamic but that there are specific properties we know exist..

Reply:
we have no rule that enforces top-level field names. So this is more like additionalProperties, with exception of kind/apiVersion/metadata.

Reply:
I don't think I can tell the CEL compiler that I want certain fields on an object to have specific types and that any others should be considered additional properties (@TristonianJones correct me if I'm wrong here).  I can do this though:

- Restrict runtime access to only the metadata name and metadata generateName fields
- Use DynType for the embedded, which still allows `embedded.spec.field` type field selection.

This isn't perfect. A CEL program like `embedded.metadata.labels.exists(...)` would compile (ideally it would fail compilation with a 'undefined field: labels' error), but this would always result in a 'no such key' error for labels at runtime.

Reply:
I've tried to address this. See https://github.com/kubernetes/kubernetes/pull/105477#discussion_r743054690

Reply:
I think I supported a set of structured properties with an `additionalProperties` field could, in theory, refer to `map(string, dyn)` or maybe just `dyn`.  Though maybe I removed support for this since I didn't think this was a good experience. You'd just have to modify the `TypeProvider` to support `additionalProperties` on a well-defined object type, so that the `FindFieldType` call would fallback to whatever type is indicated in `additionalProperties`